runtime/src/bank.rs

```diff
@@ -6534,14 +6544,58 @@ impl Bank {
         new_feature_activations: &HashSet<Pubkey>,
     ) {
         for builtin in BUILTINS.iter() {
+            // The `builtin_is_bpf` flag is used to handle the case where a
+            // builtin is scheduled to be enabled by one feature gate and
+            // later migrated to Core BPF by another.
+            //
+            // There should never be a case where a builtin is set to be
+            // migrated to Core BPF and is also set to be enabled on feature
+            // activation on the same feature gate. However, the
+            // `builtin_is_bpf` flag will handle this case as well, electing
+            // to first attempt the migration to Core BPF.
+            //
+            // The migration to Core BPF will fail gracefully because the
+            // program account will not exist. The builtin will subsequently
+            // be enabled, but it will never be migrated to Core BPF.
+            //
+            // Using the same feature gate for both enabling and migrating a
+            // builtin to Core BPF should be strictly avoided.
+            let mut builtin_is_bpf = false;
+            if let Some(core_bpf_migration_config) = &builtin.core_bpf_migration_config {
+                // If the builtin is set to be migrated to Core BPF on feature
+                // activation, perform the migration and do not add the program
+                // to the bank's builtins. The migration will remove it from
+                // the builtins list and the cache.
+                if new_feature_activations.contains(&core_bpf_migration_config.feature_id) {
+                    if let Err(e) = self
+                        .migrate_builtin_to_core_bpf(&builtin.program_id, core_bpf_migration_config)
+                    {
+                        warn!(
+                            "Failed to migrate builtin {} to Core BPF: {}",
+                            builtin.name, e
+                        );
+                    } else {
+                        builtin_is_bpf = true;
+                    }
+                } else {
+                    // If the builtin has already been migrated to Core BPF, do not
+                    // add it to the bank's builtins.
+                    builtin_is_bpf = self
+                        .get_account(&builtin.program_id)
+                        .map(|a| a.owner() == &bpf_loader_upgradeable::id())
+                        .unwrap_or(false);
+                }
+            };
+
```

Comment:
I think it would be better to insert into / remove from `new_feature_activations` instead of fiddling with `should_apply_action_for_feature_transition`.

Reply:
I'm not sure I follow what you mean here. Could you elaborate?

Reply:
`compute_active_feature_set()` figures out what is active / inactive while `apply_feature_activations()` and `apply_builtin_program_feature_transitions()` perform the transition.

Reply:
Right I understand that, I'm just wondering what you mean by "fiddling with `should_apply_action_for_feature_transition`".

The original check for `should_apply_action_for_feature_transition` is still in-tact, but it also checks to see if the program is already BPF, and I've renamed it since it now corresponds only to activation. Previously "enabling" a program was the only possible action, but now we have two: enabling and migrating (disabling).

These are two separate feature gates. Does that make sense or am I missing something?

Reply:
@Lichtso how else might you envision this check, as per your previous comments?

Reply:
We could suppress it in `new_feature_activations`